### ISSUE TITLE
Special treatment of `rule all`

### ISSUE THREAD

**Issue by SultanOrazbayev** [snakemake/snakemake#23]
**Is your feature request related to a problem? Please describe.**

By default, Snakemake will execute the first rule in a workflow. Current `rule dependency` syntax can only be used sequentially. If one wants to express `rule all` with rule-dependent inputs, then `all` must be placed at the end and explicitly requested when running `snakemake` (there are other hacks around this).

**Describe the solution you'd like**

It would be great if `rule all` is treated as an exception, so that if `rule all` is defined (anywhere in the workflow), then it is treated as the default target rule.

**Comment by elsherbini** [comment]
I personally dislike the "rule all" pattern. Often it's not "all", you can have targets defined by some rules that don't get called by the topmost pseudo rule. I wish the tutorials and documentation would change "all" to "default_target" or something else unambiguous. 

The name doesn't really matter of course, but as soon as there is a special case for some names I think it'd be worth thinking about the choice more carefully.

Another solution that doesn't mean making special cases for some names is to treat all pseudo rules the same way - if you are a rule that doesn't have an output, automatically get evaulated after all other rules are instantiated (still top to bottom?).

**Comment by SultanOrazbayev** [comment]
re: treating all rules the same way.
It's possible that some rules are a utility (e.g. deleting specific files/folders), so activating them automatically might be problematic. However, if a rule could have a Boolean property like `evaluate_manually`, then this could work.

**Comment by elsherbini** [comment]
I don't mean to actually execute the rule, just evaluate its inputs so you could use the `rule.whatever.output` pattern in the input of that rule, but still have it above the others.

**Comment by troycomi** [comment]
I don't think this is going to get picked up because snakemake is following the conventions of make, which are entrenched. 
https://www.gnu.org/software/make/manual/html_node/Rules.html
> The order of rules is not significant, except for determining the default goal: the target for make to consider, if you do not otherwise specify one. The default goal is the target of the first rule in the first makefile.

and 

> Therefore, we usually write the makefile so that the first rule is the one for compiling the entire program or all the programs described by the makefile (often with a target called ‘all’).

So you are probably stuck calling `snakemake all`.  More fundamentally, snakemake has to infer from the output file wildcards to eventually begin running the workflow.  Consider:
```
rule all:
    input: rule.dummy.output

rule dummy:
    output: {sample}.txt
    shell: 'touch {output}'
```
Forcing the default rule to have an output independent of other rules ensures that cases like the above can resolve.

Personally I like keeping all my paths in a dictionary in the config file and then utilize those names directly:
```
configfile: config.yaml  # paths: dummy: {sample}.txt
paths = config['paths']
rule all:
    input: paths['dummy'].format(sample='test')

rule dummy:
    output: paths['dummy']
    shell: 'touch {output}'
```
You get similar semantics to using rule dependency while expressing the filenames.  It's also easier to customize all the outputs in the config file instead of hunting through the snakefile.

**Comment by SultanOrazbayev** [comment]
@troycomi Thanks, this is an interesting approach. Is there a way to somehow add the rule dependency in the config file?

In your example, if there's another rule, `dummy2`, that uses input from `dummy`, then one would need to specify the path twice (especially if each rule has multiple named outputs and only some of them are needed in a dependent rule). The code below doesn't work, but is there a way to use your approach to get it to work:
```
paths:
  dummy:
    out: '{sample}.txt'
  dummy2:
    in: rules.dummy.output.out
```

**Comment by troycomi** [comment]
You could, but generally it seems cleaner to define your rule dependencies in the Snakefile alone and the paths in the config.  Using variables in the yaml file is supported in some versions of yaml, but I usually write my own utility script to replace instances of `'__DUMMY__'` with the value of dummy so I don't have to mess with the yaml library snakemake uses.

A more real world example may be useful here as the issues you raise don't really come up.  Consider the workflow of indexing a bam then running haplotype caller on it.  The yaml would be:
```
paths:
  bam: 'path/to/input/bams/{sample}.bam'
  bai: 'path/to/input/bams/{sample}.bai'
  gvcf: 'where/to/store/results/{sample}.g.vcf'
  gvcf_idx: 'where/to/store/results/{sample}.g.vcf.idx'
```
The Snakefile is then
```
rule index_bam:
  input: paths['bam']
  output: paths['bai']
  shell: ...

rule haplotype_caller:
  input:
    bam=paths['bam'],
    bai=paths['bai']
  output:
    gvcf=paths['gvcf'],
    gvcf_idx=paths['gvcf_idx']
  shell: ...
```

For larger workflows, I find the rule referencing to be hard to follow, especially if the rules are split into subworkflows.  Duplicating lines isn't as bad since the path dictionary is shorter than writing `rules.index_bam.output[0]` and all the paths are in a central location (the config).

In practice, I use the string replacement I mentioned above to have the config be closer to:
```
paths:
  input_base: 'path/to/input/bams'
  output_base: 'where/to/store/results'

  bam: '__INPUT_BASE__/{sample}.bam'
  bai: '__INPUT_BASE__/{sample}.bai'
  gvcf: '__OUTPUT_BASE__/{sample}.g.vcf'
  gvcf_idx: '__OUTPUT_BASE__/{sample}.g.vcf.idx'
```
So the user only has to change the base files once prior to rerunning the workflow.  I know setting the working directory can achieve the same thing, but for some reason I like having it in the config still.

And really it's personal preference.  I find the conifg reference to be easier to use as it's independent of rule order and more descriptive than using rule references.

**Comment by SultanOrazbayev** [comment]
Thank you very much, @troycomi!

**Comment by cademirch** [comment]
This was solved with addition of default_target directive
